### PR TITLE
Updates Dawn to use the new structured_data filter

### PR DESCRIPTION
**Pull request by lhoffbeck**
### PR Summary: 

Replace product and article structured data generated in the theme with structured data generated by the `structured_data` filter.

### Why are these changes introduced?

Updating themes to stay on the greenpath of structured data developments is currently pretty painful. 

### Decision log

| # | Decision | Alternatives | Rationale | Downsides |
|---|---|---|---|---|
| 1 |   |   |   |   |

### Testing steps/scenarios

- [ ] Step 1

### Demo links

- [Store](url)
- [Editor](url)

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

**Comment by bakura10** [comment]
That's a very nice update. A few notes though:

* The schema used in Dawn is incomplete. I've maintained one over the years, that also include some extra things that would be nice to have: https://gist.github.com/bakura10/7704584dc866d7ae3754789a6fdc66b4
* It would be better to have support for blog post ({{ article | structured_data }}), and one for shop ({{ shop | structured_data }}) that would output this: https://gist.github.com/bakura10/7704584dc866d7ae3754789a6fdc66b4#file-microdata-schema-liquid-L234

**Comment by lhoffbeck** [comment]
> May I suggest this be implemented as a containerless `{{ structured_data }}` added to `layout/theme.liquid`?
> 
> This would allow Shopify to contextually include Organization, WebSite, Article, BreadcrumbList, FAQPage, anything else that might be useful, and adapt to best practices in future.

That's great feedback, I'll take this back to the team

**Comment by tairli** [comment]
What would be the recommended way to integrate product reviews?
It was popular to add the data from metafields into the ld+json element like in @bakura10 gist.

**Comment by lhoffbeck** [comment]
@bakura10 yeah good idea, I have it in a backlog ticket. Both/and improvement along with this :)

> What would be the recommended way to integrate product reviews?
> It was popular to add the data from metafields into the ld+json element like in @bakura10 gist.

@tairli great question! Shopify-generated data only includes information that does not require any inference. Since reviews tend to be stored as metafields and don't have a standard definition, they don't fall under this case. However, the Shopify data can be trivially [extended](https://developers.google.com/search/docs/appearance/structured-data/sd-policies#multiple-items) by adding another block with the same `@id` to the page. For example:

```html
<!-- data from {{ product | structured_data }} -->
<script type="application/ld+json">
{
   "@context":"http://schema.org/",
   "@id":"/products/my-product#product",
   "name":"My product",
   ... other fields ...
}
</script>

<!-- @tairli's enhanced data -->
<script type="application/ld+json">
{
   "@context":"http://schema.org/",
   "@id":"/products/my-product#product",
   "aggregateRating": {
        "@type": "AggregateRating",
        "ratingValue": "5",
        "reviewCount": "100"
   }
}
</script>
```

**Comment by bakura10** [comment]
Makes sense, but I think this will cause us a lot of support :/. Not all review apps actually output this, so merchants will see the missing review and will just come back to us on how to add it. It would definitely be needed for merchants to have control over that.

**Comment by lhoffbeck** [comment]
Thanks @joshistoast ! This should support 3p reviews apps already, the flow would look something like this:
- Review app sets metafields on the product
- Merchant's theme outputs product structured data using the filter
- App injects structured data for the review into the storefront, with an `"@id": "{{ product.url }}#product"`. Google combines the app block with the structured_data filter block by the `@id` prop.

App dev guidance to the merchant would be something like: if you use the structured_data filter, review app structured data works automatically. If you customize structured_data, either ensure an `"@id": "{{ product.url }}#product"` is present or include this review block: `{"aggregateRating": ...}`

> would it be worth exploring an argument expecting a specific reviews structure so we can format and use those

Can you tell me more about what you're thinking here?

**Comment by joshistoast** [comment]
> Can you tell me more about what you're thinking here?

I'm more-so curious to overriding values in case the shape of data is odd or id I's like to change it in some way, in which case I'm sure the traditional structured data approach would be best.

@lhoffbeck

File: sections/main-article.liquid
Comment (lhoffbeck) [inline sections/main-article.liquid:1]: I removed the property-based SEO from this file. Google interprets this as a duplicate item when present along with ld+json: https://search.google.com/test/rich-results/result?id=6ClESn2Nb85_l_LNc1XbVA

**Review by tyleralsbury — APPROVED** [review]
Tested.  Loving the net `-76` line diff, too. :chefs-kiss:

**Review by dan-menard — APPROVED** [review]

**Comment by haroldao** [comment]
image
An error is returned in the article template...

**Comment by lhoffbeck** [comment]
Can you drop a link so I can take a look?

**Comment by haroldao** [comment]
Yeah sure,

https://harold-theme-dev.myshopify.com/blogs/news/article-2?_ab=0&_bt=eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaEpJaU5vWVhKdmJHUXRkR2hsYldVdFpHVjJMbTE1YzJodmNHbG1lUzVqYjIwR09nWkZWQT09IiwiZXhwIjoiMjAyNC0wNS0zMFQxNjo1Njo1OC44ODVaIiwicHVyIjoicGVybWFuZW50X3Bhc3N3b3JkX2J5cGFzcyJ9fQ%3D%3D--ba89f82cc7cd0c8345fd8ed22075705877884806&_fd=0&_sc=1&key=65835c145ec874592a1d2daf3e489b6c57f36d8b8b42bc41f07f8cb06a019304&preview_theme_id=137161244838

**Comment by joshistoast** [comment]
Created a ticket with yotpo so they can start supporting this new filter 👀

**Comment by bakura10** [comment]
aaaaand as expected, support debt started after we upgraded to the structured_data filter :( !

image

Unfortunately, Shopify structured_data does not output all the data, and this leads to warnings on Google merchants. I agree that those are just warning and won't affect in any way SEO ranking, but merchants tend do be super picky about those stuff and it is hard for us to explain that to merchants, as they see the warning and expect it to be fixed.

As we now use the structured_data we are going to redirect all merchants to Shopify support, but still, this create some support debt on our end that we would like to avoid.

For reference, please find again my final implementation of the structured data: https://gist.github.com/bakura10/7704584dc866d7ae3754789a6fdc66b4

@lhoffbeck can you please escalate this issue and ensure that this field (and in general all fields triggering a warning for which you have the info ; which is the case here) can be part of the schema outputted by the filter?

**Comment by lhoffbeck** [comment]
Hey Michael, we should be able to add that field :)

**Comment by bakura10** [comment]
Thanks, please let me know once added so I can let the merchant know! :)

**Comment by lhoffbeck** [comment]
Hey @bakura10 , `hasMerchantReturnPolicy` should be supported as part of the filter now.

**Comment by bakura10** [comment]
Thanks! I can confirm thsi is here now!

**Comment by Misfi** [comment]
@lhoffbeck are there any plans to support more properties, and fully support existing ones?
E.g. you mentioned `hasMerchantReturnPolicy`, but it's not yet everywhere:
{48C1F671-64FA-4735-A58E-A753388B3915}

**Pull request merged**